### PR TITLE
Backport: Unified snapshots configuration for datanode 

### PR DESCRIPTION
**Pull request by todvora**
Backport of #21031

fixes https://github.com/Graylog2/graylog2-server/issues/20859

### CODE REVIEW

**Comment by moesterheld** [comment]
Since we have a lot of users setting up data node in docker environments where the watermark issue is triggered, @sethgraylog requested a backport to 6.1

File: changelog/unreleased/pr-21031.toml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+type = "a"
+message = "Unified snapshot configuration, role and cache management in datanode"
```
Comment (moesterheld) [inline changelog/unreleased/pr-21031.toml:2]:
```suggestion
message = "Unified snapshot configuration, role and cache management in datanode. This prevents allocating a big search cache when no searchable snapshots have been configured, leading to potential low watermark issues."
```

**Review by moesterheld — CHANGES_REQUESTED** [review]
the pr is missing some changes for the keystore cli that were introduced, so it isn't working right now

```
2024-12-03 10:38:04,281 ERROR: org.graylog.datanode.opensearch.statemachine.OpensearchStateMachine - Failed to execute opensearch cli/Users/oesterheld/dev/projects/graylog-project-master/graylog-project-repos/graylog2-server/data-node/target/opensearch/opensearch-2.15.0-linux-aarch64/bin/opensearch-keystore
STDOUT/STDERR: Enter value for plugins.security.ssl.transport.keystore_password_secure: Exception in thread "main" java.lang.NullPointerException: Cannot read the array length because "array" is null
	at java.base/java.nio.CharBuffer.wrap(CharBuffer.java:468)
	at org.opensearch.common.settings.KeyStoreWrapper.setString(KeyStoreWrapper.java:638)
	at org.opensearch.common.settings.AddStringKeyStoreCommand.executeCommand(AddStringKeyStoreCommand.java:117)
	at org.opensearch.common.settings.BaseKeyStoreCommand.execute(BaseKeyStoreCommand.java:89)
	at org.opensearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:104)
	at org.opensearch.cli.Command.mainWithoutErrorHandling(Command.java:138)
	at org.opensearch.cli.MultiCommand.execute(MultiCommand.java:104)
	at org.opensearch.cli.Command.mainWithoutErrorHandling(Command.java:138)
	at org.opensearch.cli.Command.main(Command.java:101)
	at org.opensearch.common.settings.KeyStoreCli.main(KeyStoreCli.java:56)
```

**Comment by todvora** [comment]
> the pr is missing some changes for the keystore cli that were introduced, so it isn't working right now
> 
> ```
> 2024-12-03 10:38:04,281 ERROR: org.graylog.datanode.opensearch.statemachine.OpensearchStateMachine - Failed to execute opensearch cli/Users/oesterheld/dev/projects/graylog-project-master/graylog-project-repos/graylog2-server/data-node/target/opensearch/opensearch-2.15.0-linux-aarch64/bin/opensearch-keystore
> STDOUT/STDERR: Enter value for plugins.security.ssl.transport.keystore_password_secure: Exception in thread "main" java.lang.NullPointerException: Cannot read the array length because "array" is null
> 	at java.base/java.nio.CharBuffer.wrap(CharBuffer.java:468)
> 	at org.opensearch.common.settings.KeyStoreWrapper.setString(KeyStoreWrapper.java:638)
> 	at org.opensearch.common.settings.AddStringKeyStoreCommand.executeCommand(AddStringKeyStoreCommand.java:117)
> 	at org.opensearch.common.settings.BaseKeyStoreCommand.execute(BaseKeyStoreCommand.java:89)
> 	at org.opensearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:104)
> 	at org.opensearch.cli.Command.mainWithoutErrorHandling(Command.java:138)
> 	at org.opensearch.cli.MultiCommand.execute(MultiCommand.java:104)
> 	at org.opensearch.cli.Command.mainWithoutErrorHandling(Command.java:138)
> 	at org.opensearch.cli.Command.main(Command.java:101)
> 	at org.opensearch.common.settings.KeyStoreCli.main(KeyStoreCli.java:56)
> ```

I think what we are missing is [this undocumented -x argument](https://github.com/Graylog2/graylog2-server/pull/20923/files#diff-c7b479ce8013f91d1fd4b80e5696463af50c00473c0371251a30772951e9d302R45) from another PR. All our tests are green, so I assume it's only a mac problem and the -x would be good enough and simple solution, easy to backport.

I still think it makes sense to backport this, otherwise we'll be dealing with watermark issues and support tickets till the next release.

**Review by moesterheld — APPROVED** [review]
re-tested everything. thanks for the backport!

**Pull request merged**